### PR TITLE
Fix region nonfungible implementation

### PR DESCRIPTION
**Pull request by Szegoo**
The problem with the current implementation is that minting will cause the region coremask to be set to `Coremask::complete` regardless of the actual coremask.

This PR fixes that.

More details about the nonfungible implementation can be found here: https://github.com/paritytech/polkadot-sdk/pull/3455

### CODE REVIEW

**Review by seadanda — COMMENTED** [review]
Oof, good catch. Would be nice to add a test to interlace or partition->burn->mint and then check the mask is identical after minting. This PR should also be backported to v1.14

**Comment by Szegoo** [comment]
> Would be nice to add a test to interlace or partition->burn->mint and then check the mask is identical after minting.

I added two new tests. They don't explicitly check for the coremask, but if the coremask changed when minting, the region returned would be the old one which still has the owner set to `None`, so the test would fail without this fix.

**Comment by seadanda** [comment]
Looks good, just needs a prdoc. I just realised it's breaking as `issue` is pub, so it might be an issue to backport.

**Comment by Szegoo** [comment]
> I just realised it's breaking as `issue` is pub, so it might be an issue to backport.

But it is not exposed as an extrinsic; it can only be accessed from the runtime code, right?

File: substrate/frame/broker/src/utility_impls.rs [hunk 1]
```diff
@@ -94,11 +94,12 @@ impl<T: Config> Pallet<T> {
 	pub fn issue(
```
Comment (seadanda) [inline substrate/frame/broker/src/utility_impls.rs:94]: The more I think about this the more I think this should just be `pub(crate)` - I see no reason for this to be exposed as `pub`
```suggestion
	pub(crate) fn issue(
```

File: prdoc/pr_5067.prdoc [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+      the region coremask to be set to `Coremask::complete` regardless of the
+      actual coremask of the region.
+
+crates: 
+- name: pallet-broker
+  bump: major
+- name: coretime-rococo-runtime
+  bump: major
+- name: coretime-westend-runtime
+  bump: major
```
Comment (seadanda) [inline prdoc/pr_5067.prdoc:19]:
```suggestion
- name: coretime-rococo-runtime
  bump: patch
- name: coretime-westend-runtime
  bump: patch
```

I don't think this is any more than a bugfix for the runtimes

File: substrate/frame/broker/src/utility_impls.rs (hunk 1 above)
Comment (seadanda) [reply substrate/frame/broker/src/utility_impls.rs:94]: no wait, ignore this

File: prdoc/pr_5067.prdoc (hunk 2 above)
Comment (Szegoo) [reply prdoc/pr_5067.prdoc:19]: True, since we don't expose the `issue` function.

Comment (Szegoo) [reply prdoc/pr_5067.prdoc:19]: True, since we don't expose the `issue` function.

**Comment by ggwpez** [comment]
The semver CI check should be fixed no master, please try again after merging it in.

**Review by seadanda — APPROVED** [review]

File: cumulus/parachains/runtimes/coretime/coretime-rococo/src/lib.rs [hunk 3]
```diff
@@ -51,6 +51,8 @@ use frame_system::{
 	limits::{BlockLength, BlockWeights},
 	EnsureRoot,
 };
+#[cfg(feature = "runtime-benchmarks")]
+use pallet_broker::CoreMask;
```
Comment (bkontur) [inline cumulus/parachains/runtimes/coretime/coretime-rococo/src/lib.rs:55]:
```suggestion
```
```suggestion
#[cfg(feature = "runtime-benchmarks")]
use pallet_broker::CoreMask;
```

File: cumulus/parachains/runtimes/coretime/coretime-rococo/src/lib.rs [hunk 4]
```diff
@@ -832,7 +834,7 @@ impl_runtime_apis! {
 					let begin = 0;
 					let end = 42;
 
-					let region_id = pallet_broker::Pallet::<Runtime>::issue(core, begin, end, None, None);
+					let region_id = pallet_broker::Pallet::<Runtime>::issue(core, begin, CoreMask::complete(), end, None, None);
```
Comment (bkontur) [inline cumulus/parachains/runtimes/coretime/coretime-rococo/src/lib.rs:837]:
```suggestion
					let region_id = pallet_broker::Pallet::<Runtime>::issue(core, begin, pallet_broker::CoreMask::complete(), end, None, None);
```

File: cumulus/parachains/runtimes/coretime/coretime-westend/src/lib.rs [hunk 5]
```diff
@@ -51,6 +51,8 @@ use frame_system::{
 	limits::{BlockLength, BlockWeights},
 	EnsureRoot,
 };
+#[cfg(feature = "runtime-benchmarks")]
+use pallet_broker::CoreMask;
```
Comment (bkontur) [inline cumulus/parachains/runtimes/coretime/coretime-westend/src/lib.rs:55]:
```suggestion
```

File: cumulus/parachains/runtimes/coretime/coretime-westend/src/lib.rs [hunk 6]
```diff
@@ -823,7 +825,7 @@ impl_runtime_apis! {
 					let begin = 0;
 					let end = 42;
 
-					let region_id = pallet_broker::Pallet::<Runtime>::issue(core, begin, end, None, None);
+					let region_id = pallet_broker::Pallet::<Runtime>::issue(core, begin, CoreMask::complete(), end, None, None);
```
Comment (bkontur) [inline cumulus/parachains/runtimes/coretime/coretime-westend/src/lib.rs:828]:
```suggestion
					let region_id = pallet_broker::Pallet::<Runtime>::issue(core, begin, pallet_broker::CoreMask::complete(), end, None, None);
```

**Review by bkontur — APPROVED** [review]
lgtm, just nits with import (can be also ignored)

**Comment by Szegoo** [comment]
The semver check was failing. Maybe merging with the master will fix it?

**Review by franciscoaguirre — APPROVED** [review]

**Pull request merged**